sae_lens/saes/sae_base.py

```diff
@@ -0,0 +1,588 @@
+                raise ValueError("TopK activation function requires a k value.")
+            k = kwargs.get("k", 1)  # Default k to 1 if not provided
+
+            def topk_fn(x: torch.Tensor) -> torch.Tensor:
+                topk = torch.topk(x.flatten(start_dim=-1), k=k, dim=-1)
+                values = torch.relu(topk.values)
+                result = torch.zeros_like(x.flatten(start_dim=-1))
+                result.scatter_(-1, topk.indices, values)
+                return result.view_as(x)
+
+            return topk_fn
+        raise ValueError(f"Unknown activation function: {activation_fn}")
+
+    def _setup_activation_normalization(self):
+        """Set up activation normalization functions based on config."""
+        if self.cfg.normalize_activations == "constant_norm_rescale":
+
+            def run_time_activation_norm_fn_in(x: torch.Tensor) -> torch.Tensor:
+                self.x_norm_coeff = (self.cfg.d_in**0.5) / x.norm(dim=-1, keepdim=True)
+                return x * self.x_norm_coeff
+
+            def run_time_activation_norm_fn_out(x: torch.Tensor) -> torch.Tensor:
+                x = x / self.x_norm_coeff  # type: ignore
+                del self.x_norm_coeff
+                return x
+
+            self.run_time_activation_norm_fn_in = run_time_activation_norm_fn_in
+            self.run_time_activation_norm_fn_out = run_time_activation_norm_fn_out
+
+        elif self.cfg.normalize_activations == "layer_norm":
+
+            def run_time_activation_ln_in(
+                x: torch.Tensor, eps: float = 1e-5
+            ) -> torch.Tensor:
+                mu = x.mean(dim=-1, keepdim=True)
+                x = x - mu
+                std = x.std(dim=-1, keepdim=True)
+                x = x / (std + eps)
+                self.ln_mu = mu
+                self.ln_std = std
+                return x
+
+            def run_time_activation_ln_out(
+                x: torch.Tensor,
+                eps: float = 1e-5,  # noqa: ARG001
+            ) -> torch.Tensor:
+                return x * self.ln_std + self.ln_mu  # type: ignore
+
+            self.run_time_activation_norm_fn_in = run_time_activation_ln_in
+            self.run_time_activation_norm_fn_out = run_time_activation_ln_out
+        else:
+            self.run_time_activation_norm_fn_in = lambda x: x
+            self.run_time_activation_norm_fn_out = lambda x: x
+
+    @abstractmethod
+    def initialize_weights(self):
+        """Initialize model weights."""
+        pass
+
+    @abstractmethod
+    def encode(
+        self, x: Float[torch.Tensor, "... d_in"]
+    ) -> Float[torch.Tensor, "... d_sae"]:
+        """Encode input tensor to feature space."""
+        pass
+
+    @abstractmethod
+    def decode(
+        self, feature_acts: Float[torch.Tensor, "... d_sae"]
+    ) -> Float[torch.Tensor, "... d_in"]:
+        """Decode feature activations back to input space."""
+        pass
+
+    def turn_on_forward_pass_hook_z_reshaping(self):
```

Comment:
It seems like we have a lot of stuff that's in the category of "do stuff to activations between they reach the main SAE methods". This class of things feels like they can be moved into their own `ActivationPreprocessor` classes or something, and then the SAE can just run them all to preprocess its input. This would be things like reshaping, layer norm, scaling by the activation norm, etc... These could have methods like `process_input` and `process_output` if they need to do processing on the output as well (e.g. scale inputs before they get to the SAE, then unscale the outputs again). These could be passed as a list in the SAE config maybe. Doesn't necessarily need to be part of this PR, but maybe worth discussing more.